### ISSUE TITLE
Due dates for GitHub Classroom 

### ISSUE THREAD

**Issue by mozzadrella** [education/classroom#959]
## Overview
6 respondents
3 teachers who are Classroom users
1 teacher who does not use Classroom
2 testers from usertesting.com

tl;dr 
* surface the SHA + state at the deadline (and make clear that’s what it is, that it’s the SHA at deadline, not submission)
* also surface the latest SHA, even after the deadline, most instructors want to see that
* the most important number is if there are zero commits (especially for group work)
* teachers are concerned about the amount of time from when student first accepts to due date (and activity during that time)

## How do you use due dates for your assignments?

Respondent A: Doesn’t open all assignments at once—does it one at a time. Expects students to accept it when it’s announced, but wants more insight into activity after they accept it.

Respondent B: Gives assignments with multiple parts—students who are on the ball can finish the assignment and move on to part 2. I’ve had to ask students “which is your part 1 submission?” 

Respondent C: I put them in Canvas. However, students know they can submit a link to their repo in Canvas and keep working on it. 

Respondent F: Has a google calendar with “soft deadlines.” Has scripts that check for the last push. Student loses 10% each day after the deadline. 

## What do you absolutely need to know when you check due dates? (Prioritization of info)

Respondent A:
* In teams: wants to see in teams how many commits (zero is the important number here) 
* Individual: activity beyond submission—>accepted when/last worked on it when

Respondent B:
* 3 pieces of data: 
    * Who got it in
    * who got it in, but late and 
    * who hasn’t finished 
* Students will submit after the deadline, he will want to see commits after the deadline 
* Insight into consistent effort: tick marks from the time of submission to deadline, wants to see students working semi-regularly

Respondent C: I want to know which work was submitted before the due date, and what was submitted after. Currently I give full credit for work submitted after the due date. However, I’m considering changing this to a cap (max score of 4 out of 5) next year.

## What should deadlines absolutely do? (Functionality)
Respondent B: Would want to be able to change the deadline, extension is either class-wide or individual 
Respondent C: Would want to see the state of the project at the deadline, but would also assess work after.
Respondent F: The last push is all I care about. I also want to know if they’ve done any work at all. 

## Interface observations
Make due date:
* Respondents B and C blew right past the field—should we move the calendar from a modal into the flow?
* Setting a due date was smooth and fast.

Check due date:
* Icon is too small, but well-located
* Respondents B and C know to click there, but not exactly what to expect (in current UI, submission + URL for repo are awfully close) 

## What would be nice-to-have?
* Respondent B: would like to be able to set a difficulty level for assignment
* Respondent F: number of days passed since the deadline, four hours. In a perfect world, there would be a counter or a context in the repository (can we do that with milestones?)
* Respondent B: a temperature bar to show commits/progress
* Respondent B: Tick marks from the time of submission to deadline, wants to see students working semi-regularly
* Respondent B: Sometimes the assignment is too hard—for whatever reason the class as a whole is too hard 

Other key data:
Respondent B: When I work in time management, it’s usually another tool (Jira, etc)

**Comment by mozzadrella** [comment]
Follow up from Respondent C:
>This isn’t from our conversation, but I thought of another use case. At times, students complete the assignment early yet are hesitant to explore extensions since they fear breaking their lab. While I reassure them this is why we use source control, if there was a way for a student to mark a given commit (perhaps a tag) as their submission, it would make it easier for me to rollback to this earlier version if needed. As I type this, I could just have students create a tag and that would work fine….

**Comment by nwoodthorpe** [comment]
Thanks so much for doing this survey :)

So looks like the biggest concerns/feature requests are:

1) Make it more clear that viewing a submission is viewing a commit SHA
2) Have a clear, obvious way to view the student's latest work, not only their submitted work.
3) Show the number of commits, perhaps highlight repos with 0 commits
4) Summary of repository activity
5) Sort repos by (submitted, not submitted)
6) Introduce a concept of late submissions?
7) Make due-date field more obvious
8) Be able to "submit" a commit that isn't the latest commit

@mozzadrella does that sound right? Am I missing anything important?

My work in #920 solves points 2 and 3. I need to do a bit more on that now that deadlines are shipped, but that should be good to go super soon.

Looking at the other points:

1) This sounds like a messaging problem. If our current implementation isn't as intuitive as we thought, we might need extra documentation. Maybe some tooltips explaining certain things, or link to a page that describes deadlines somewhere?

4) This is a hard problem, but I think it'd be great to implement. GitHub has some really cool graphs on a repository under insights/graphs. Maybe we can find inspiration there. There isn't much room on the assignment overview page, so I'm imagining either a tooltip the teacher can hover over to view history for users, or a separate page we link to from the assignment overview page that shows activity graphs for each repository. Another idea would be to have a button on the assignment repo view that says "View activity" and when clicked, it replaces the whole repo partial with a graph. There can be a button on that partial to return back to the old one. If you're confused by what I'm saying here, I can outline it with pictures :P

5) I think this is a great idea. It'll have to be a full page reload because of our pagination, but I don't see why this can't be done. How would this play with multiple deadlines?

6) I don't know how to solve this under our current model right now, I'll think about it a bit more. Right now, the student can't really "submit" something, so the idea of "submitting" 2 days late isn't really possible.

7) This is something we can definitely do. I'll fiddle around with some designs soon.

@tarebyte @johndbritton

**Comment by mozzadrella** [comment]
>Make it more clear that viewing a submission is viewing a commit SHA
>Have a clear, obvious way to view the student's latest work, not only their submitted work.
>Show the number of commits, perhaps highlight repos with 0 commits

@nwoodthorpe think these are within grasp.

**Comment by nwoodthorpe** [comment]
@mozzadrella 

So now that our UI changes have shipped, I think we've addressed the main concerns. This is the new UI:

<img width="1072" alt="screen shot 2017-07-12 at 3 16 41 pm" src="https://user-images.githubusercontent.com/4149056/28142334-42279ff0-6715-11e7-8b91-35bb1d1a5d3e.png">

- We now show the number of commits
- We have a "View Repository" button that always brings the teacher to the latest version of the repo
- When hovering over the "submitted" text, we see "View submission".

Were there any other main concerns before we full ship this? I'll fix the default time for deadlines today.

**Comment by mozzadrella** [comment]
@nwoodthorpe I think it looks great--can't wait to share with teachers widely!

**Comment by nwoodthorpe** [comment]
Going to close this, I think we've addressed the action items we pulled from this research. Thanks Vanessa!
